include/ginkgo/core/base/lin_op.hpp

```diff
@@ -562,6 +562,112 @@ class Permutable {
 };
 
 
+/**
+ * The RowGatherable is an interface that allows to get the row gather result of
+ * a LinOp.
+ *
+ * @tparam IndexType  the type of the indices used to extract/gather the rows
+ */
+template <typename IndexType>
+class RowGatherable {
+public:
+    /**
+     * Returns the LinOp containing a copy of the gathered rows
+     *
+     * @param gather_indeces  pointer to an array containing row indices
+     *                        from this matrix. It may contain duplicates.
+     *
+     * @return a pointer to the new LinOp containing a copy of the gathered rows
+     */
+    virtual std::unique_ptr<LinOp> row_gather_linop(
+        const Array<IndexType>* gather_indices) const = 0;
+
+    /**
+     * Copies the given rows from this matrix into `row_collection`
+     *
+     * @param gather_indices  pointer to an array containing row indices
+     *                        from this matrix. It may contain duplicates.
+     * @param row_collection  pointer to a LinOp that will store the gathered
+     *                        rows:
+     *                        `row_collection(i,j)
+     *                         = input(gather_indices(i), j)`
+     *                        It must have the same number of columns as this
+     *                        matrix and `gather_indices->get_num_elems()` rows.
+     */
+    virtual void row_gather(const Array<IndexType>* gather_indices,
+                            LinOp* row_collection) const = 0;
+
+
+    /**
+     * Copies the given rows from this matrix into `row_collection` with scaling
+     *
+     * @param alpha  scaling the result of row gathering
+     * @param gather_indices  pointer to an array containing row indices
+     *                        from this matrix. It may contain duplicates.
+     * @param beta  scaling the input row_collection
+     * @param row_collection  pointer to a LinOp that will store the
+     *             gathered rows:
+     *             `row_collection(i,j) = input(gather_indices(i), j)`
+     *             It must have the same number of columns as this
+     *             matrix and `gather_indices->get_num_elems()` rows.
+     */
+    virtual void row_gather(const LinOp* alpha,
+                            const Array<IndexType>* gather_indices,
+                            const LinOp* beta, LinOp* row_collection) const = 0;
+
+    virtual ~RowGatherable() = default;
+};
+
+
+/**
+ * The EnableRowGather mixin provides the default implementations of
+ * `row_gather_linop` and the row_gather interface.
+ *
+ * @tparam OutputType  the OutputType which is being returned
+ *                     [CRTP parameter]
+ * @tparam IndexType  the type of the indices used to extract/gather the rows
+ *
+ * @ingroup LinOp
+ */
+template <typename OutputType, typename IndexType>
+class EnableRowGather : public RowGatherable<IndexType> {
+public:
+    using output_type = OutputType;
+    // void row_gather(const Array<IndexType>*, LinOp*) const;
```

Comment:
Commented code

Reply:
no, it describes the following `using ..` brings which code to avoid compilation issue.
I add more sentences to describe it